src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -438,25 +505,69 @@ HeapWord* ShenandoahFreeSet::try_allocate_in(ShenandoahHeapRegion* r, Shenandoah
       size_t waste = r->free();
       if (waste > 0) {
         increase_used(waste);
-        generation->increase_allocated(waste);
+	generation->increase_allocated(waste);
         _heap->notify_mutator_alloc_words(waste >> LogHeapWordSize, true);
       }
+    } else if (r->free() < PLAB::min_size() * HeapWordSize) {
+      // Permanently retire this region if there's room for a fill object
+      if (r->free() >= ShenandoahHeap::min_fill_size()) {
+        size_t waste = r->free();
+        HeapWord* fill_addr = r->top();
+        size_t fill_size = waste / HeapWordSize;
+        ShenandoahHeap::fill_with_object(fill_addr, fill_size);
+        r->set_top(r->end());
+	// Since we have filled the waste with an empty object, account for increased usage
+	_heap->increase_used(waste);
+	if (_heap->mode()->is_generational()) {
+	  _heap->generation_for(req.affiliation())->increase_used(waste);
+          if (req.affiliation() == ShenandoahRegionAffiliation::OLD_GENERATION) {
+            _heap->card_scan()->register_object(fill_addr);
```

Comment:
I'm asking myself whether this needs a lock.  This code is inside try_allocate_in(), which is called by allocate_single() and allocate_with_affiliation() and allocate_old_with_affiliation(), which are called by allocate(), which is called by ShenandoahHeap::allocate_memory_under_lock();  so I have the lock.

Other threads may be concurrently registering objects that belong to the same region without acquring a lock, but this would only happen for objects allocated from within a plab, and we know that none of those will touch the same memory as is affected by this operation, so no locking is necessary.